arrow-check/src/main/kotlin/arrow/check/property/Impl.kt

```diff
@@ -0,0 +1,389 @@
+package arrow.check.property
+
+import arrow.check.FailureAnnotation
+import arrow.check.FailureSummary
+import arrow.check.Progress
+import arrow.check.Report
+import arrow.check.gen.Gen
+import arrow.check.gen.RandSeed
+import arrow.check.gen.Rose
+import arrow.check.gen.flatMap
+import arrow.check.gen.map
+import arrow.check.gen.runEnv
+import arrow.check.testCount
+import arrow.core.Tuple3
+import kotlinx.atomicfu.atomic
+import kotlinx.coroutines.flow.collect
+import pretty.Doc
+import pretty.doc
+import pretty.hardLine
+import pretty.plus
+import pretty.spaced
+import pretty.text
+import kotlin.coroutines.Continuation
+import kotlin.coroutines.coroutineContext
+import kotlin.coroutines.intrinsics.COROUTINE_SUSPENDED
+import kotlin.coroutines.intrinsics.startCoroutineUninterceptedOrReturn
+import kotlin.coroutines.intrinsics.suspendCoroutineUninterceptedOrReturn
+import kotlin.coroutines.resume
+
+internal data class State(
+  val numTests: TestCount,
+  val numDiscards: DiscardCount,
+  val size: Size,
+  val seed: RandSeed,
+  val coverage: Coverage<CoverCount>
+)
+
+// ---------------- Running a single property
+internal class PropertyTestImpl : PropertyTest {
+
+  var promise: Promise<Gen<Any?, TestResult<*>>> = Promise()
+
+  suspend fun waitFor(): Gen<Any?, TestResult<*>> = promise.await()
+    .also { promise = Promise() }
+
+  fun completeOrFlatMap(gen: Gen<Any?, TestResult<*>>): Unit = promise.complete(gen) { new, curr ->
+    curr.flatMap { currRes ->
+      new.map { newRes -> newRes.prependLog(currRes.log) }
+    }
+  }.let { }
+
+  override suspend fun <R, A> forAllWith(showA: (A) -> Doc<Markup>, env: R, gen: Gen<R, A>): A =
+    suspendCoroutineUninterceptedOrReturn { c ->
+      val labelHere = c.stateStack // save the whole coroutine stack labels
+      val interceptedC = intercepted.get(c)
+      var first = false
+      gen.runEnv(env).flatMap { x: A ->
+        c.stateStack = labelHere
```

Comment:
The only reasonable alternative is to force users to use just one `forAll` per test and combine all generators with `tupledN` or offer an api that does it for them. `forAll` could also function like `flatMap` but that opens up other pitfalls that a) lead to nesting and b) lead to easy to make mistakes.
```kotlin
check {
  val a = forAll(genA)
  val b = forAll(genB)
  val c = forAll(genC)
  ...
}
```
vs
```kotlin
check(tupledN(genA, genB, genC)) { (a, b, c) ->
  ...
}
```

The flat forAll using the suspend hack is by far the nicest syntax, if only it would work on mpp...

Reply:
> The flat forAll using the suspend hack is by far the nicest syntax, if only it would work on mpp...

100% agreed. The flat syntax is much nicer and more powerful. Although `tupledN` could be hidden from the users API with some boilerplate on the library side writing arity-N overloads... but yes that's not amazing compared to the flat syntax. The flat syntax also comes with major runtime costs due to the stack label resetting.

Sadly, I think that the second option might be better for the time being.

Reply:
> The flat syntax also comes with major runtime costs due to the stack label resetting

This only happens on shrinking (well atm it does for both, but by checking `first` before a reset it could), so normal execution where most tests are expected to succeed don't really get hurt from this. Also when a test fails it usually (depending on the test and the shrinker) shrinks in the low 100s of iterations and the printing, diffing and reporting easily dwarves that cost in those cases. Also lets not talk about performance in arrow-check, it's probably fast enough but there are a lot of cases to be improved :see_no_evil: 

Actually since generators are monadic and allow suspend the power of the api is the same if we have one or many `forAll`'s, its just simply not nearly as nice to use...

So lots do a quick pro and cons:
Flat `forAll`:
Pro:
 - **Much nicer syntax**

Cons:
 - **No (full) mpp** (it should be able to support js/jvm through expect/actual)
 - Slows down shrinking
 - More complex implementation which may be a source of errors

Top level `forAll`:
Pro:
 - **Full mpp**
 - Easier to implement
 - Faster when shrinking

Cons:
 - **Annoying to use, monadic use becomes much harder**

If there is a way to (gracefully) fallback to top level gens on native that would be great, since I really don't want to loose this syntax, but also want this to be multi-platform...